bundles/org.openhab.binding.easee/src/main/resources/OH-INF/thing/charger-channel-groups.xml

```diff
@@ -6,9 +6,13 @@
 	<channel-group-type id="charger-state">
 		<label>Charger Status</label>
 		<channels>
-			<channel id="smartCharging" typeId="type-switch">
+			<channel id="smartCharging" typeId="rwtype-switch">
```

Comment:
By changing the channel type, I assume you have now to provide migration stuff (something new in OH4) for a transparent migration for users that created their things with UI.

Reply:
A link to the documentation
https://next.openhab.org/docs/developer/bindings/thing-xml.html#updating-thing-types

Reply:
Thanks for the hint. I did not know about this feature. As I use file based configuration I always recreated stuff. But that is quite inconvenient if you created your config via UI.

Reply:
This is a very new and interesting feature of OH 4, apparently not yet fully stable (I tested it yesterday for the first time). 
It is a good point that binding developers consider it now.

Reply:
added migration. documentation is quite basic I had to take a look into XSD to find out how to adress channel groups. Not sure if this is correct now.

Reply:
This is probably correct but we would expect that you first test it.

Please note that PR openhab/openhab-core#3576 which improves the thing upgrade. I will merge your PR only after the merge of the mentioned core PR. And then you could try again the result of your thing upgrade ?

Reply:
I will try this. So test scenario is that I will start with openhab 3.4.0 and then upgrade to 4.0.0 and the expected result is that the thing is updated without manual interaction (such as removing and readding it). Correct?

Reply:
You just need to have the binding running before this PR with the corresponding thing created using MainUI, whatever it is OH 3 or OH 4.
Then install this new version and check that your thing is properly upgraded. 

I would suggest you wait for a fixed version of OH4 first.

Reply:
I guess that my migration is not complete. I just changed the type but I guess I also need to add the properties.

However testing fails due to some other technical issue once I rebase on latest main I cannot start the binding anymore

```
19:59:27.573 [ERROR] [core.thing.internal.ThingManagerImpl] - bundle org.openhab.core.thing:4.0.0.202305080909 (161)[org.openhab.core.thing.internal.ThingManagerImpl(227)] : The addThingHandlerFactory method has thrown an exception
java.lang.NoSuchMethodError: 'java.lang.Object javax.xml.bind.ServiceLoaderUtil.lookupUsingOSGiServiceLoader(java.lang.String, java.util.logging.Logger)'
	at javax.xml.bind.ContextFinder.find(ContextFinder.java:413) ~[?:?]
	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:721) ~[?:?]
	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:662) ~[?:?]
	at org.openhab.core.thing.internal.update.ThingUpdateInstructionReaderImpl.readForFactory(ThingUpdateInstructionReaderImpl.java:84) ~[?:?]
	at org.openhab.core.thing.internal.ThingManagerImpl.addThingHandlerFactory(ThingManagerImpl.java:1142) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invokeMethod(BaseMethod.java:245) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.access$500(BaseMethod.java:41) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod$Resolved.invoke(BaseMethod.java:687) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BaseMethod.invoke(BaseMethod.java:531) ~[?:?]
	at org.apache.felix.scr.impl.inject.methods.BindMethod.invoke(BindMethod.java:42) ~[?:?]
	at org.apache.felix.scr.impl.manager.DependencyManager.doInvokeBindMethod(DependencyManager.java:2086) ~[?:?]
	at org.apache.felix.scr.impl.manager.DependencyManager.invokeBindMethod(DependencyManager.java:2061) ~[?:?]
	at org.apache.felix.scr.impl.manager.SingleComponentManager.invokeBindMethod(SingleComponentManager.java:443) ~[?:?]
	at org.apache.felix.scr.impl.manager.DependencyManager$MultipleDynamicCustomizer.addedService(DependencyManager.java:336) ~[?:?]
	at org.apache.felix.scr.impl.manager.DependencyManager$MultipleDynamicCustomizer.addedService(DependencyManager.java:304) ~[?:?]
	at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerAdded(ServiceTracker.java:1232) ~[?:?]
	at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.customizerAdded(ServiceTracker.java:1152) ~[?:?]
	at org.apache.felix.scr.impl.manager.ServiceTracker$AbstractTracked.trackAdding(ServiceTracker.java:959) ~[?:?]
	at org.apache.felix.scr.impl.manager.ServiceTracker$AbstractTracked.track(ServiceTracker.java:895) ~[?:?]
	at org.apache.felix.scr.impl.manager.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:1184) ~[?:?]
	at org.apache.felix.scr.impl.BundleComponentActivator$ListenerInfo.serviceChanged(BundleComponentActivator.java:116) ~[?:?]
	at org.eclipse.osgi.internal.serviceregistry.FilteredServiceListener.serviceChanged(FilteredServiceListener.java:120) ~[org.eclipse.osgi-3.17.200.jar:?]
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:956) ~[?:?]
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234) ~[?:?]
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:151) ~[?:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEventPrivileged(ServiceRegistry.java:936) ~[org.eclipse.osgi-3.17.200.jar:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.publishServiceEvent(ServiceRegistry.java:873) ~[org.eclipse.osgi-3.17.200.jar:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistrationImpl.register(ServiceRegistrationImpl.java:141) ~[org.eclipse.osgi-3.17.200.jar:?]
	at org.eclipse.osgi.internal.serviceregistry.ServiceRegistry.registerService(ServiceRegistry.java:261) ~[org.eclipse.osgi-3.17.200.jar:?]
	at org.eclipse.osgi.internal.framework.BundleContextImpl.registerService(BundleContextImpl.java:495) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:929) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager$3.register(AbstractComponentManager.java:915) ~[?:?]
	at org.apache.felix.scr.impl.manager.RegistrationManager.changeRegistration(RegistrationManager.java:133) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.registerService(AbstractComponentManager.java:984) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:752) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enableInternal(AbstractComponentManager.java:674) ~[?:?]
	at org.apache.felix.scr.impl.manager.AbstractComponentManager.enable(AbstractComponentManager.java:437) ~[?:?]
	at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.enableComponents(ConfigurableComponentHolder.java:671) ~[?:?]
	at org.apache.felix.scr.impl.BundleComponentActivator.initialEnable(BundleComponentActivator.java:310) ~[?:?]
	at org.apache.felix.scr.impl.Activator.loadComponents(Activator.java:593) ~[?:?]
	at org.apache.felix.scr.impl.Activator.access$200(Activator.java:74) ~[?:?]
	at org.apache.felix.scr.impl.Activator$ScrExtension.start(Activator.java:460) ~[?:?]
	at org.apache.felix.scr.impl.AbstractExtender.createExtension(AbstractExtender.java:196) ~[?:?]
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:169) ~[?:?]
	at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:49) ~[?:?]
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:488) ~[org.eclipse.osgi-3.17.200.jar:?]
	at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:1) ~[org.eclipse.osgi-3.17.200.jar:?]
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:232) ~[org.eclipse.osgi-3.17.200.jar:?]
	at org.osgi.util.tracker.BundleTracker$Tracked.bundleChanged(BundleTracker.java:450) ~[org.eclipse.osgi-3.17.200.jar:?]
	at org.eclipse.osgi.internal.framework.BundleContextImpl.dispatchEvent(BundleContextImpl.java:944) ~[?:?]
	at org.eclipse.osgi.framework.eventmgr.EventManager.dispatchEvent(EventManager.java:234) ~[?:?]
	at org.eclipse.osgi.framework.eventmgr.ListenerQueue.dispatchEventSynchronous(ListenerQueue.java:151) ~[?:?]
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEventPrivileged(EquinoxEventPublisher.java:229) ~[?:?]
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:138) ~[?:?]
	at org.eclipse.osgi.internal.framework.EquinoxEventPublisher.publishBundleEvent(EquinoxEventPublisher.java:130) ~[?:?]
	at org.eclipse.osgi.internal.framework.EquinoxContainerAdaptor.publishModuleEvent(EquinoxContainerAdaptor.java:217) ~[?:?]
	at org.eclipse.osgi.container.Module.publishEvent(Module.java:499) ~[?:?]
	at org.eclipse.osgi.container.Module.start(Module.java:486) ~[?:?]
	at org.eclipse.osgi.internal.framework.EquinoxBundle.start(EquinoxBundle.java:445) ~[?:?]
	at aQute.launcher.Launcher.start(Launcher.java:699) ~[?:?]
	at aQute.launcher.Launcher.startBundles(Launcher.java:679) ~[?:?]
	at aQute.launcher.Launcher.activate(Launcher.java:585) ~[?:?]
	at aQute.launcher.Launcher.launch(Launcher.java:404) ~[?:?]
	at aQute.launcher.Launcher.run(Launcher.java:186) ~[?:?]
	at aQute.launcher.Launcher.main(Launcher.java:162) ~[?:?]
	at aQute.launcher.pre.EmbeddedLauncher.executeWithRunPath(EmbeddedLauncher.java:170) ~[biz.aQute.launcher.pre.jar:?]
	at aQute.launcher.pre.EmbeddedLauncher.findAndExecute(EmbeddedLauncher.java:135) ~[biz.aQute.launcher.pre.jar:?]
	at aQute.launcher.pre.EmbeddedLauncher.main(EmbeddedLauncher.java:52) ~[biz.aQute.launcher.pre.jar:?]
19:59:34.813 [INFO ] [e.automation.internal.RuleEngineImpl] - Rule engine started.
```

Reply:
I pulled the latest changes from distro project which results in another error, openhab cannot even start anymore due to an error in Eclipse:

Resolution failed. Capabilities satisfying the following requirements could not be found:
    [<<INITIAL>>]
      ⇒ osgi.identity: (osgi.identity=org.eclipse.equinox.metatype)

Reply:
I managed to fix the error above. However there is now a new strange behaviour. On openhab startup the binding remains in state "NOT YET READY" for about two minutes. There are some warnings in the logs

```
18:54:17.742 [DEBUG] [g.easee.internal.EaseeHandlerFactory] - bundle org.openhab.binding.easee:4.0.0.202305100701 (169)[org.openhab.binding.easee.internal.EaseeHandlerFactory(295)] : invoked activate: activate
18:54:17.743 [DEBUG] [g.easee.internal.EaseeHandlerFactory] - bundle org.openhab.binding.easee:4.0.0.202305100701 (169)[org.openhab.binding.easee.internal.EaseeHandlerFactory(295)] : Set implementation object for component
18:54:17.743 [DEBUG] [g.easee.internal.EaseeHandlerFactory] - bundle org.openhab.binding.easee:4.0.0.202305100701 (169)[org.openhab.binding.easee.internal.EaseeHandlerFactory(295)] : Changed state from satisfied to active
18:54:22.578 [INFO ] [e.automation.internal.RuleEngineImpl] - Rule engine started.
18:56:18.169 [WARN ] [core.thing.internal.ThingManagerImpl] - A thing handler factory claims to support 'easee:mastercharger' for thing 'easee:mastercharger:bf43d5d35b:EHNBJJCG' for more than 120s, but the thing type can't be found in the registry. This should be fixed in the binding.
18:56:18.169 [WARN ] [core.thing.internal.ThingManagerImpl] - A thing handler factory claims to support 'easee:site' for thing 'easee:site:bf43d5d35b' for more than 120s, but the thing type can't be found in the registry. This should be fixed in the binding.
18:56:18.169 [WARN ] [core.thing.internal.ThingManagerImpl] - Could not normalize configuration for 'easee:site:bf43d5d35b' because the thing type was not found in registry.
18:56:18.179 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'easee:site:bf43d5d35b' changed from UNINITIALIZED (NOT_YET_READY) to INITIALIZING
```
Afterwards the bindings starts but the migration was not executed.

Reply:
Strange. The message means that the thing type easee:site is not found by the core framework. 
What OH version arr you running?

Reply:
I tried both

1) running it from Eclipse (main branch) with latest snapshot
2) running it standalone with latest 4.0.0-SNAPSHOT Docker Image

Result is the same. It also happens with any other binding (e.g. nibeuplink) that I add as a "custom user build" addon also if there is no change at all.

Reply:
There seems to be an issue with bindings built from Eclipse. Since recent changes in openhab core there is a warning that the maven lifecycle is not fully covered. This seems to have some impact on the result as the jar is only 80k compared to 96k when building from console.

Reply:
> I managed to fix the error above. However there is now a new strange behaviour. On openhab startup the binding remains in state "NOT YET READY" for about two minutes. There are some warnings in the logs
> 
> ```
> 18:54:17.742 [DEBUG] [g.easee.internal.EaseeHandlerFactory] - bundle org.openhab.binding.easee:4.0.0.202305100701 (169)[org.openhab.binding.easee.internal.EaseeHandlerFactory(295)] : invoked activate: activate
> 18:54:17.743 [DEBUG] [g.easee.internal.EaseeHandlerFactory] - bundle org.openhab.binding.easee:4.0.0.202305100701 (169)[org.openhab.binding.easee.internal.EaseeHandlerFactory(295)] : Set implementation object for component
> 18:54:17.743 [DEBUG] [g.easee.internal.EaseeHandlerFactory] - bundle org.openhab.binding.easee:4.0.0.202305100701 (169)[org.openhab.binding.easee.internal.EaseeHandlerFactory(295)] : Changed state from satisfied to active
> 18:54:22.578 [INFO ] [e.automation.internal.RuleEngineImpl] - Rule engine started.
> 18:56:18.169 [WARN ] [core.thing.internal.ThingManagerImpl] - A thing handler factory claims to support 'easee:mastercharger' for thing 'easee:mastercharger:bf43d5d35b:EHNBJJCG' for more than 120s, but the thing type can't be found in the registry. This should be fixed in the binding.
> 18:56:18.169 [WARN ] [core.thing.internal.ThingManagerImpl] - A thing handler factory claims to support 'easee:site' for thing 'easee:site:bf43d5d35b' for more than 120s, but the thing type can't be found in the registry. This should be fixed in the binding.
> 18:56:18.169 [WARN ] [core.thing.internal.ThingManagerImpl] - Could not normalize configuration for 'easee:site:bf43d5d35b' because the thing type was not found in registry.
> 18:56:18.179 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'easee:site:bf43d5d35b' changed from UNINITIALIZED (NOT_YET_READY) to INITIALIZING
> ```
> 
> Afterwards the bindings starts but the migration was not executed.

It seems that this was caused by a corrupt Eclipse. I created a new instance from scatch which is now working fine. However the exception is back:

```
19:59:27.573 [ERROR] [core.thing.internal.ThingManagerImpl] - bundle org.openhab.core.thing:4.0.0.202305080909 (161)[org.openhab.core.thing.internal.ThingManagerImpl(227)] : The addThingHandlerFactory method has thrown an exception
java.lang.NoSuchMethodError: 'java.lang.Object javax.xml.bind.ServiceLoaderUtil.lookupUsingOSGiServiceLoader(java.lang.String, java.util.logging.Logger)'
	at javax.xml.bind.ContextFinder.find(ContextFinder.java:413) ~[?:?]
	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:721) ~[?:?]
	at javax.xml.bind.JAXBContext.newInstance(JAXBContext.java:662) ~[?:?]
	at org.openhab.core.thing.internal.update.ThingUpdateInstructionReaderImpl.readForFactory(ThingUpdateInstructionReaderImpl.java:84) ~[?:?]
	at org.openhab.core.thing.internal.ThingManagerImpl.addThingHandlerFactory(ThingManagerImpl.java:1142) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
...
...
```
It happens as soon as I have a "instructions.xml" file in my update folder. Wehen I delete the file everything is fine and no exception occurs. Looks like this is closely related to the migration feature. @lolodomo Any hint on this?

Reply:
I worked around the Eclipse issue by testing the binding in a standalone openhab instance.

```
2023-05-11 17:16:26.449 [INFO ] [org.openhab.core.Activator          ] - Starting openHAB 4.0.0 (build Build #3458)
2023-05-11 17:16:42.841 [WARN ] [org.openhab.core.net.NetUtil        ] - Found multiple local interfaces - ignoring 172.17.0.1
2023-05-11 17:16:42.846 [WARN ] [org.openhab.core.net.NetUtil        ] - Found multiple local interfaces - ignoring 10.11.2.22
2023-05-11 17:16:49.738 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'mapdb.persist'
2023-05-11 17:16:50.639 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'influxdb.persist'
2023-05-11 17:16:52.239 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'default.sitemap'
2023-05-11 17:16:53.230 [INFO ] [el.core.internal.ModelRepositoryImpl] - Loading model 'easee.things'
2023-05-11 17:17:04.618 [INFO ] [.core.model.lsp.internal.ModelServer] - Started Language Server Protocol (LSP) service on port 5007
2023-05-11 17:17:14.343 [INFO ] [hab.io.metrics.MetricsRestController] - Metrics service activated, serving the following URL(s): /rest/metrics/prometheus
2023-05-11 17:17:15.207 [INFO ] [e.automation.internal.RuleEngineImpl] - Rule engine started.
2023-05-11 17:17:16.106 [INFO ] [.influxdb.InfluxDBPersistenceService] - InfluxDB persistence service started.
2023-05-11 17:17:20.163 [INFO ] [core.thing.internal.ThingManagerImpl] - Updating 'easee:mastercharger:ec3df6a644XXX:EHNBJJCG' from version 0 to 1
```
Migration is executed and channel type is changed.
However it is not enough to change the channel type I also need to add the properties. It seems that this is not yet supported by the migration engine. At least I cannot find any hint in documentation or XSD how to achieve this.

Reply:
Do you mean the new channel properties "writeCommand" and "validationExpression" ?
To be honest with you, I have no idea what could be the interest to define properties for channels.
When you create the thing with MainUI, you see these properties in your file org.openhab.core.thing.Thing.json ?
I tend to think that the channel properties should be populated with the properties from the channel type during migration. @J-N-K : WDYT ? Did you consider the channel properties ?

Reply:
I use the properties to store additional settings. Maybe this is not the original intention of those properties but it works for me. To have less types I decided to store additional configuration in properties:
- Command Mapping for write access
- Validation expression to validate input before sending it to the cloud API

I could of course add additional channel types e.g. a dedicated "rwtype-switch-XYZ" for each API endpoint. That way I could remove the configuration and instead hardwire each type to exactly one write API endpoint.

Reply:
By the way, it is an offered possibility so it should be handled properly by the upgrade tool.

Reply:
@alexf2015 : please declare an issue in the core repo about forgotten channel properties.

Reply:
just created https://github.com/openhab/openhab-core/issues/3610